File: mac/Stepwright/Sources/FormatProfile.swift
import Foundation

/// How a guide is turned into markup. Every system that receives a document wants something
/// slightly different, so the rules live in a file that can be edited, shared and swapped
/// rather than being buried in the exporter.
///
/// The shape matches the Windows build exactly, so a format file works on either.
struct FormatProfile: Codable {
    var Name: String = "Untitled format"
    var Description: String = ""

    // Shape
    var InlineStyles: Bool = false
    var SingleContainer: Bool = false
    var UseOrderedList: Bool = true
    var UseHeadingTags: Bool = true
    var AllowColor: Bool = true

    var IncludeTitle: Bool = true
    var IncludeSummary: Bool = true
    var IncludeMeta: Bool = true
    var IncludeFooter: Bool = true

    // Type
    var FontFamily: String = "-apple-system, system-ui, Helvetica, Arial, sans-serif"
    var TitleSize: Int = 32
    var HeadingSize: Int = 20
    var BodySize: Int = 17
    var NoteSize: Int = 15
    var MetaSize: Int = 13
    var BlockSpacing: Int = 12

    // Steps
    var StepPrefix: String = ""
    var StepSuffix: String = "."
    var BoldStepText: Bool = false
    var NotePrefix: String = ""

    // Pictures
    var ImageWidth: Int = 1400
    var UseJpeg: Bool = false
    var JpegQuality: Double = 82
    var AllowAnimation: Bool = true
    var EmbedImages: Bool = true
    var ImageDisplayWidth: Int = 0
    var RoundImageCorners: Bool = true

    /// Written in place of each picture. The number of the step replaces {n}.
    var ImagePlaceholder: String = ""

    // Extras
    var FooterText: String = "Made with Stepwright"
    var Preamble: String = ""

    var isBuiltIn: Bool = false

    enum CodingKeys: String, CodingKey {
        case Name, Description, InlineStyles, SingleContainer, UseOrderedList, UseHeadingTags
        case AllowColor, IncludeTitle, IncludeSummary, IncludeMeta, IncludeFooter
        case FontFamily, TitleSize, HeadingSize, BodySize, NoteSize, MetaSize, BlockSpacing
        case StepPrefix, StepSuffix, BoldStepText, NotePrefix
        case ImageWidth, UseJpeg, JpegQuality, AllowAnimation, EmbedImages
        case ImageDisplayWidth, RoundImageCorners, ImagePlaceholder
        case FooterText, Preamble
    }
}

/// The formats that ship with the app, and the folder holding any others.
enum FormatProfiles {
    static let fileExtension = "swformat"

    static var folder: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return base.appendingPathComponent("Stepwright/formats")
    }

    /// The look the app uses unless told otherwise.
    static func standard() -> FormatProfile {
        var profile = FormatProfile()
        profile.Name = "Stepwright"
        profile.Description = "The look the app uses by default: a styled page with rounded pictures."
        profile.isBuiltIn = true
        profile.SingleContainer = true
        profile.BoldStepText = true
        return profile
    }

    /// What Hudu accepts: one container, inline styles, Arial, sixteen point bold headings,
    /// fourteen point body, no tables and no colour so the site controls light and dark mode.
    static func hudu() -> FormatProfile {
        var profile = FormatProfile()
        profile.Name = "Hudu"
        profile.Description = "One container, inline styles, Arial, no colour so Hudu controls light and dark mode."
        profile.isBuiltIn = true
        profile.InlineStyles = true
        profile.SingleContainer = true
        profile.UseHeadingTags = false
        profile.AllowColor = false
        profile.IncludeMeta = false
        profile.FontFamily = "Arial, sans-serif"
        profile.TitleSize = 16
        profile.HeadingSize = 16
        profile.BodySize = 14
        profile.NoteSize = 14
        profile.MetaSize = 12
        profile.BoldStepText = false
        profile.NotePrefix = "Note: "
        profile.UseJpeg = true
        profile.JpegQuality = 78
        profile.ImageWidth = 1100
        profile.ImageDisplayWidth = 700
        profile.RoundImageCorners = false
        profile.AllowAnimation = false
        profile.FooterText = "Published from Stepwright on {date}"
        return profile
    }

    /// Confluence keeps pictures as attachments and refers to them by name, so the markup
    /// carries a reference rather than the picture itself.
    static func confluence() -> FormatProfile {
        var profile = FormatProfile()
        profile.Name = "Confluence"
        profile.Description = "Storage format, with pictures attached to the page and referred to by name."
        profile.isBuiltIn = true
        profile.UseOrderedList = false
        profile.AllowColor = false
        profile.FontFamily = ""
        profile.BoldStepText = true
        profile.UseJpeg = true
        profile.JpegQuality = 80
        profile.ImageWidth = 1100
        profile.ImageDisplayWidth = 700
        profile.AllowAnimation = false
        profile.RoundImageCorners = false
        profile.ImagePlaceholder = "<ac:image ac:width=\"700\"><ri:attachment ri:filename=\"step{n}.jpg\" /></ac:image>"
        profile.FooterText = "Published from Stepwright on {date}"
        return profile
    }

    /// Markup with nothing added, for pasting into an editor that styles it itself.
    static func plain() -> FormatProfile {
        var profile = FormatProfile()
        profile.Name = "Plain"
        profile.Description = "Headings, paragraphs and pictures with no styling at all."
        profile.isBuiltIn = true
        profile.AllowColor = false
        profile.IncludeMeta = false
        profile.IncludeFooter = false
        profile.FontFamily = ""
        profile.BoldStepText = true
        profile.RoundImageCorners = false
        return profile
    }

    static func builtIn() -> [FormatProfile] { [standard(), hudu(), confluence(), plain()] }

    /// Everything that ships with the app plus everything saved beside it.
    static func all() -> [FormatProfile] {
        var profiles = builtIn()

        guard let files = try? FileManager.default.contentsOfDirectory(
            at: folder,
            includingPropertiesForKeys: nil) else { return profiles }

        for file in files where file.pathExtension == fileExtension {
            guard let loaded = load(file) else { continue }

            // A saved format with the same name replaces the one that ships with the app.
            profiles.removeAll { $0.Name.caseInsensitiveCompare(loaded.Name) == .orderedSame }
            profiles.append(loaded)
        }

        return profiles
    }

    static func find(_ name: String?) -> FormatProfile {
        let profiles = all()
        guard let name else { return profiles[0] }
        return profiles.first { $0.Name.caseInsensitiveCompare(name) == .orderedSame } ?? profiles[0]
    }

    static func load(_ url: URL) -> FormatProfile? {
        guard let data = try? Data(contentsOf: url),
              var profile = try? JSONDecoder().decode(FormatProfile.self, from: data),
              !profile.Name.isEmpty else { return nil }

        profile.isBuiltIn = false
        return profile
    }

    static func save(_ profile: FormatProfile) {
        try? FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        try? write(profile, to: folder.appendingPathComponent(safeName(profile.Name) + "." + fileExtension))
    }

    static func write(_ profile: FormatProfile, to url: URL) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try encoder.encode(profile).write(to: url)
    }

    static func delete(_ name: String) {
        let url = folder.appendingPathComponent(safeName(name) + "." + fileExtension)
        try? FileManager.default.removeItem(at: url)
    }

    private static func safeName(_ name: String) -> String {
        let clean = name
            .replacingOccurrences(of: "/", with: " ")
            .replacingOccurrences(of: ":", with: " ")
            .trimmingCharacters(in: .whitespaces)

        return clean.isEmpty ? "format" : clean
    }
}

File: src/Stepwright/Model/FormatProfile.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Stepwright.Model;

/// <summary>
/// How a guide is turned into markup. Every system that receives a document wants something
/// slightly different, so the rules live in a file that can be edited, shared and swapped
/// rather than being buried in the exporter.
/// </summary>
public sealed class FormatProfile
{
    /// <summary>Name shown in the list. Also the file name when exported.</summary>
    public string Name { get; set; } = "Untitled format";

    public string Description { get; set; } = string.Empty;

    // ---------------------------------------------------------------- shape

    /// <summary>Style rules travel on each element rather than in a stylesheet.</summary>
    public bool InlineStyles { get; set; }

    /// <summary>Everything sits inside one element, which some systems insist on.</summary>
    public bool SingleContainer { get; set; }

    /// <summary>Numbered steps use a list rather than a numbered heading per step.</summary>
    public bool UseOrderedList { get; set; } = true;

    /// <summary>Heading levels are written as h2 and h3 rather than styled text.</summary>
    public bool UseHeadingTags { get; set; } = true;

    /// <summary>Colour is left to the receiving system, for light and dark mode.</summary>
    public bool AllowColor { get; set; } = true;

    public bool IncludeTitle { get; set; } = true;
    public bool IncludeSummary { get; set; } = true;
    public bool IncludeMeta { get; set; } = true;
    public bool IncludeFooter { get; set; } = true;

    // ---------------------------------------------------------------- type

    public string FontFamily { get; set; } = "Segoe UI, system-ui, Helvetica, Arial, sans-serif";
    public int TitleSize { get; set; } = 32;
    public int HeadingSize { get; set; } = 20;
    public int BodySize { get; set; } = 17;
    public int NoteSize { get; set; } = 15;
    public int MetaSize { get; set; } = 13;
    public int BlockSpacing { get; set; } = 12;

    // ---------------------------------------------------------------- steps

    /// <summary>Wording placed before the number, for example "Step ".</summary>
    public string StepPrefix { get; set; } = string.Empty;

    /// <summary>What follows the number when steps are not written as a list.</summary>
    public string StepSuffix { get; set; } = ".";

    public bool BoldStepText { get; set; }

    /// <summary>Wording placed before a note, for example "Note: ".</summary>
    public string NotePrefix { get; set; } = string.Empty;

    // ---------------------------------------------------------------- pictures

    public int ImageWidth { get; set; } = 1400;

    /// <summary>Pictures are compressed rather than kept lossless, which keeps a page small.</summary>
    public bool UseJpeg { get; set; }

    public long JpegQuality { get; set; } = 82;

    /// <summary>Animated steps are written as animations where the format allows it.</summary>
    public bool AllowAnimation { get; set; } = true;

    /// <summary>Pictures travel inside the markup rather than beside it.</summary>
    public bool EmbedImages { get; set; } = true;

    /// <summary>Widest a picture is drawn in the finished page. Zero leaves it alone.</summary>
    public int ImageDisplayWidth { get; set; }

    public bool RoundImageCorners { get; set; } = true;

    /// <summary>
    /// Written in place of each picture instead of the picture itself. The number of the step
    /// replaces {n}. Used by systems that attach pictures separately.
    /// </summary>
    public string ImagePlaceholder { get; set; } = string.Empty;

    // ---------------------------------------------------------------- extras

    /// <summary>Added at the end when a footer is wanted. The date replaces {date}.</summary>
    public string FooterText { get; set; } = "Made with Stepwright";

    /// <summary>Written before everything else, exactly as given.</summary>
    public string Preamble { get; set; } = string.Empty;

    [JsonIgnore]
    public bool IsBuiltIn { get; set; }

    public FormatProfile Copy()
    {
        string json = JsonSerializer.Serialize(this);
        FormatProfile clone = JsonSerializer.Deserialize<FormatProfile>(json) ?? new FormatProfile();
        clone.IsBuiltIn = false;
        return clone;
    }
}

/// <summary>The formats that ship with the app, and the folder holding any others.</summary>
public static class FormatProfiles
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNameCaseInsensitive = true,
    };

    public const string FileExtension = ".swformat";
    public const string FileFilter = "Stepwright format (*.swformat)|*.swformat|JSON file (*.json)|*.json";

    public static string Folder =>
        Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "Stepwright",
            "formats");

    /// <summary>The look the app uses unless told otherwise.</summary>
    public static FormatProfile Standard() => new()
    {
        Name = "Stepwright",
        Description = "The look the app uses by default: a styled page with rounded pictures.",
        IsBuiltIn = true,
        InlineStyles = false,
        SingleContainer = true,
        UseOrderedList = true,
        UseHeadingTags = true,
        AllowColor = true,
        BoldStepText = true,
        RoundImageCorners = true,
    };

    /// <summary>
    /// What Hudu accepts: one container, inline styles, Arial, sixteen point bold headings,
    /// fourteen point body, no tables and no colour so the site controls light and dark mode.
    /// </summary>
    public static FormatProfile Hudu() => new()
    {
        Name = "Hudu",
        Description = "One container, inline styles, Arial, no colour so Hudu controls light and dark mode.",
        IsBuiltIn = true,
        InlineStyles = true,
        SingleContainer = true,
        UseOrderedList = true,
        UseHeadingTags = false,
        AllowColor = false,
        IncludeMeta = false,
        FontFamily = "Arial, sans-serif",
        TitleSize = 16,
        HeadingSize = 16,
        BodySize = 14,
        NoteSize = 14,
        MetaSize = 12,
        BlockSpacing = 12,
        BoldStepText = false,
        NotePrefix = "Note: ",
        UseJpeg = true,
        JpegQuality = 78,
        ImageWidth = 1100,
        ImageDisplayWidth = 700,
        RoundImageCorners = false,
        AllowAnimation = false,
        FooterText = "Published from Stepwright on {date}",
    };

    /// <summary>
    /// Confluence keeps pictures as attachments and refers to them by name, so the markup
    /// carries a reference rather than the picture itself.
    /// </summary>
    public static FormatProfile Confluence() => new()
    {
        Name = "Confluence",
        Description = "Storage format, with pictures attached to the page and referred to by name.",
        IsBuiltIn = true,
        InlineStyles = false,
        SingleContainer = false,
        UseOrderedList = false,
        UseHeadingTags = true,
        AllowColor = false,
        IncludeMeta = true,
        FontFamily = string.Empty,
        BoldStepText = true,
        StepPrefix = string.Empty,
        StepSuffix = ".",
        UseJpeg = true,
        JpegQuality = 80,
        ImageWidth = 1100,
        ImageDisplayWidth = 700,
        AllowAnimation = false,
        RoundImageCorners = false,
        ImagePlaceholder = "<ac:image ac:width=\"700\"><ri:attachment ri:filename=\"step{n}.jpg\" /></ac:image>",
        FooterText = "Published from Stepwright on {date}",
    };

    /// <summary>Markup with nothing added, for pasting into an editor that styles it itself.</summary>
    public static FormatProfile Plain() => new()
    {
        Name = "Plain",
        Description = "Headings, paragraphs and pictures with no styling at all.",
        IsBuiltIn = true,
        InlineStyles = false,
        SingleContainer = false,
        UseOrderedList = true,
        UseHeadingTags = true,
        AllowColor = false,
        IncludeMeta = false,
        IncludeFooter = false,
        FontFamily = string.Empty,
        BoldStepText = true,
        RoundImageCorners = false,
    };

    public static List<FormatProfile> BuiltIn() => new() { Standard(), Hudu(), Confluence(), Plain() };

    /// <summary>Everything that ships with the app plus everything saved beside it.</summary>
    public static List<FormatProfile> All()
    {
        List<FormatProfile> profiles = BuiltIn();

        try
        {
            if (!Directory.Exists(Folder))
            {
                return profiles;
            }

            foreach (string file in Directory.GetFiles(Folder, "*" + FileExtension))
            {
                FormatProfile? loaded = Load(file);
                if (loaded is null)
                {
                    continue;
                }

                // A saved format with the same name replaces the one that ships with the app.
                profiles.RemoveAll(p => p.Name.Equals(loaded.Name, StringComparison.OrdinalIgnoreCase));
                profiles.Add(loaded);
            }
        }
        catch
        {
            // A damaged folder should never stop the app from exporting.
        }

        return profiles;
    }

    public static FormatProfile Find(string? name)
    {
        List<FormatProfile> profiles = All();
        return profiles.FirstOrDefault(p => p.Name.Equals(name, StringComparison.OrdinalIgnoreCase))
            ?? profiles[0];
    }

    public static FormatProfile? Load(string path)
    {
        try
        {
            FormatProfile? profile = JsonSerializer.Deserialize<FormatProfile>(
                File.ReadAllText(path),
                JsonOptions);

            if (profile is null || string.IsNullOrWhiteSpace(profile.Name))
            {
                return null;
            }

            return profile;
        }
        catch
        {
            return null;
        }
    }

    public static void Save(FormatProfile profile)
    {
        Directory.CreateDirectory(Folder);
        File.WriteAllText(Path.Combine(Folder, SafeName(profile.Name) + FileExtension), Write(profile));
    }

    public static void Export(FormatProfile profile, string path) => File.WriteAllText(path, Write(profile));

    public static string Write(FormatProfile profile) => JsonSerializer.Serialize(profile, JsonOptions);

    public static void Delete(string name)
    {
        try
        {
            string path = Path.Combine(Folder, SafeName(name) + FileExtension);
            if (File.Exists(path))
            {
                File.Delete(path);
            }
        }
        catch
        {
            // Nothing useful to do if the file is locked.
        }
    }

    private static string SafeName(string name)
    {
        string clean = name;
        foreach (char bad in Path.GetInvalidFileNameChars())
        {
            clean = clean.Replace(bad, ' ');
        }

        clean = clean.Trim();
        return clean.Length == 0 ? "format" : clean;
    }
}
